### PR TITLE
[WIP] static code optimization

### PR DESCRIPTION
**Pull request by free-coffee**
#### Summary
`SUMMARY: None`
#### Purpose of change
Removing redundancies for small runtime/compile(-memory) optimizations and reduced code size.

#### Describe the solution
Using cppcheck and include-what-you-use as static analysis tools and implemented their most reasonable comments.

#### Describe alternatives you've considered
Tried to use cinclude2dot to search the most expensive headers (which are map.h (230 edges) and translations.h (200) i think) to reduce linking memory, but didn't find anything to optimize.
Also skipped the tests/*.cpp includes.

### CODE REVIEW

File: src/condition.cpp [hunk 1]
```diff
@@ -22,13 +20,16 @@
...
 #include "overmapbuffer.h"
-#include "recipe.h"
 #include "recipe_groups.h"
 #include "string_id.h"
 #include "type_id.h"
 #include "vehicle.h"
 #include "vpart_position.h"
 
+class basecamp;
+class recepie;
```
Comment (Night-Pryanik) [inline src/condition.cpp]:
```suggestion
class recipe;
```

File: src/json.cpp [hunk 2]
```diff
@@ -182,26 +182,26 @@ std::string JsonObject::str()
...
 }
 
-void JsonArray::throw_error( std::string err )
+void JsonArray::throw_error( const std::string &err )
 {
     jsin->error( err );
 }
 
-void JsonArray::throw_error( std::string err, int idx )
+void JsonArray::throw_error( const std::string &err, const int &idx )
```
Comment (Qrox) [inline src/json.cpp]: Why would a reference to an int be better here?

Comment (free-coffee) [reply src/json.cpp]: To not create a local copy if you don't need one.

File: src/vehicle.h [hunk 3]
```diff
@@ -524,7 +523,7 @@ class turret_data
 struct label : public point {
     label() = default;
     label( const point &p ) : point( p ) {}
-    label( const point &p, std::string text ) : point( p ), text( std::move( text ) ) {}
+    label( const point &p, const std::string &text ) : point( p ), text( std::move( text ) ) {}
```
Comment (Qrox) [inline src/vehicle.h]:
```suggestion
    label( const point &p, const std::string &text ) : point( p ), text( text ) {}
```

File: src/json.cpp (hunk 2 above)
Comment (Qrox) [reply src/json.cpp]: A local copy of an `int` is on the stack, while a reference is behind the scene an address that needs to be dereferenced. Given that an `int` usually has lower or equal bit width than an address, I would expect the local-copy to be faster to access?

**Comment by free-coffee** [comment]
For some reason the game startup is slower and the game crashes on loading a game which might take some time to figure out.

**Comment by kevingranade** [comment]
In future I recommend you break this kind of thing into much smaller pieces.

**Comment by free-coffee** [comment]
@kevingranade This is a small change, since if done correctly it doesn't change anything but optimize objectivly better code. The line or file-change count should matter for evaluating a change.

**Comment by kevingranade** [comment]
The line and file change counts always matter. If they don't it means you're not reviewing the changes.

**Pull request closed without merging**